TaskModules/powershell/TaskModuleIISManageUtility/AppCmdOnTargetMachines.ps1

```diff
@@ -1,525 +1,813 @@
+        [string] [Parameter(Mandatory=$true)] $virtualPath,
+        [string] [Parameter(Mandatory=$true)] $physicalPath
+    )
+
+    $tmpPhysicalPath = $physicalPath.Replace("%SystemDrive%", "$env:SystemDrive")
+    Write-Verbose -verbose "Checking website physical path exists $tmpPhysicalPath"
+    if(!(Test-Path -Path $tmpPhysicalPath))
+    {
+        Write-Verbose -verbose "Creating application physical path $tmpPhysicalPath"
+        New-Item -ItemType Directory -Path $tmpPhysicalPath
+    }
+
+    $appCmdPath, $iisVersion = Get-AppCmdLocation -regKeyPath $AppCmdRegKey
+    $appCmdArgs = [string]::Format(' add app /site.name:"{0}" /path:"/{1}" /physicalPath:"{2}"',$siteName, $virtualPath, $physicalPath)
+    $command = "`"$appCmdPath`" $appCmdArgs"
+
+    Write-Verbose -verbose "Creating web application. Running command : $command"
+    Invoke-Appcmd -command $command
+}
+
+
+function Update-Application 
+{
+    param (
+        [string] [Parameter(Mandatory=$true)] $applicationName,
+        [string] $physicalPath,
+        [string] $applicationPool,
+        [string] $physicalPathAuthentication,
+        [System.Management.Automation.PSCredential] $physicalPathAuthenticationCredentials
+    )
+
+    $appCmdArgs = [string]::Format(' set app /app.name:"{0}"', $applicationName)
+
+    if(-not [string]::IsNullOrWhiteSpace($applicationPool))
+    {
+        $appCmdArgs = [string]::Format('{0} -applicationPool:"{1}"', $appCmdArgs, $applicationPool)
+    }
+
+    if(-not [string]::IsNullOrWhiteSpace($physicalPath))
+    {
+        $tmpPhysicalPath = $physicalPath.Replace("%SystemDrive%", "$env:SystemDrive")
+        Write-Verbose -verbose "Checking application physical path exists $tmpPhysicalPath"
+        if(!(Test-Path -Path $tmpPhysicalPath))
+        {
+            Write-Verbose -verbose "Creating application physical path $tmpPhysicalPath"
+            New-Item -ItemType Directory -Path $tmpPhysicalPath
+        }
+        $appCmdArgs = [string]::Format("{0} -[path='/'].physicalPath:`"{1}`"", $appCmdArgs, $physicalPath)
+    }
+
+    if($physicalPathAuthentication -eq "ApplicationWindowsAuth") 
+    {
+        $userName = $physicalPathAuthenticationCredentials.userName
+        $password = $physicalPathAuthenticationCredentials.GetNetworkCredential().password
+
+        if(-not [string]::IsNullOrWhiteSpace($userName))
+        {
+            $appCmdArgs = [string]::Format("{0} -[path='/'].userName:{1}", $appCmdArgs, $userName)
+        }
+
+        if(-not [string]::IsNullOrWhiteSpace($password))
+        {
+            $appCmdArgs = [string]::Format("{0} -[path='/'].password:{1}", $appCmdArgs, $password)
+        }
+    }
+    else 
+    {
+        $appCmdArgs = [string]::Format("{0} -[path='/'].userName:{1} -[path='/'].password:{2}", $appCmdArgs, $null, $null)
+    }
+
+    $appCmdPath, $iisVersion = Get-AppCmdLocation -regKeyPath $AppCmdRegKey
+    $command = "`"$appCmdPath`" $appCmdArgs"

```

Comment:
$password having special characters will break. $appCmdArgs is not escaped or splatted.